### ISSUE TITLE
Excessive DNS queries

### ISSUE THREAD

**Issue by bkury** [croneter/PlexKodiConnect#1768]
When watching a video plexkodiconnect syncs the "played time" with the plex server. Unfortunately, every request (~1/sec) also does a dns lookup ( IP-ID.plex.direct ) which results in a ton of unnecessary dns lookups.

In some cases, e.g. when using pi-hole or some advanced router, the software starts blocking/limiting the source IP.

It would be great if you can fix that ...
(e.g. only do a dns lookup when starting a video and cache the dns response for subsequent requests)

Thank you!

**Comment by croneter** [comment]
I personally use pi-hole - no issue whatsoever.

PKC cannot "fix" this - you would need to tell Plex to stop using `plex.direct` URLs. Also, shouldn't DNS lookups be cached?!?

**Comment by bkury** [comment]
> PKC cannot "fix" this - you would need to tell Plex to stop using plex.direct URLs. Also, shouldn't DNS lookups be cached?!?

Can't you just lookup plex.direct once  you start a video and store the response in a python variable and then use that for the "played time" http requests? 

> I personally use pi-hole - no issue whatsoever.

That's because one PKC watch session is still within pi-hole server limits but more than that and you start having issues. Anyway, a lot of people use different dns servers directly (google/cloudfront/provider) which might cause issues and unnecessary dns traffic.

The point is, that it's just not necessary to do a dns lookup on every "played time" http update request. Just check your own pihole statistic when watching a video and you can see it yourself :

![pkc-pihole](https://user-images.githubusercontent.com/1777146/148048493-001daedd-9f73-470f-aebb-1418a972c9f1.jpg)

Thanks!

**Comment by croneter** [comment]
Do you actually have an issue here, something not working? Everytime you visit `google.com` you issue a DNS lookups. I do NOT want to fiddle with DNS, has been completely unnecessary in the past

**Comment by croneter** [comment]
And again: pi-hole caches these lookups of `plex.direct`!

**Comment by bkury** [comment]
> Do you actually have an issue here, something not working?

Yes, there are situations where you could experience some issues but that depends on certain factors like os / dns server used. 

> every time you visit google.com you issue a DNS lookup, too

Well, yes but only for the first time you visit the site, A browser caches the dns response for a certain amount of time, so when you open google again, it uses the cached dns response instead of asking the dns server again and again.

That's what I meant, PKC should also cache dns requests for a certain amount of time.

> And again: pi-hole caches these lookups of plex.direct!

Yes it does but that's not the real issue here.
What about most people who don't use a local dns but instead use google's or their providers dns server? 

Doing a dns lookup every second is just not very efficient and could potentially cause issues with different dns servers or PKC itself (e.g. on slow network connections). 

It also adds overhead and increases the response time because every second PKC needs to wait for a dns response before sending a "played time" update to the actual plex server.

Thanks!

**Comment by bkury** [comment]
While we are on this topic (just another thought) :

The official plex clients update their timeline only every 10 seconds instead of PKCs ~1 second. 
Such a PKC change would further reduce unnecessary traffic, database updates and logfile entries (if enabled)

This is nothing major but it would help in the long run and make PKC to be more in line with the official plex clients.

**Comment by croneter** [comment]
> Yes, there are situations where you could experience some issues but that depends on certain factors like os / dns server used.

Closing this feature request as I will never implement it. All highly hypothetical, no real issue here at all. Also, PlexKodiConnect is NOT responsible for caching DNS; I won't start messing with DNS. 

> The official plex clients update their timeline only every 10 seconds instead of PKCs ~1 second. Such a PKC change would further reduce unnecessary traffic, database updates and logfile entries (if enabled)

Dismissed as well, no real issue here.

**Comment by bkury** [comment]
Just to wrap this up, in case someone has similar problems:

Nothing I wrote is hypothetical and yes there are some real issues here but even if there weren't it's still not a good practice to do a dns lookup every second. 

Anyway, I finally had some spare time to look at the code and I think I found the problem:

`lib/plexbmchelper/subscribers.py (def _send_pms_notification)`

`DU().downloadUrl(url,
authenticate=False,
parameters=xargs,
headerOverride=HEADERS_PMS,
timeout=timeout)
`

Setting "authenticate=False" will cause DU().downloadUrl() to use requests.get() instead of requests.session() for all ":/timeline" requests which is the reason why every timeline update is a new http request and therefore a dns lookup. 

Simply removing "authenticate=False" (defaults to True) fixes this problem because it will then use requests.session() (keep-alive) for these kind of requests.

Because these timeline updates are only sent to the plex server, setting "authenticate=True" shouldn't be an security issue, I guess. Anyway, everything looks good now

Feel free to use this information, or not ...

**Comment by iKeule257** [comment]
Hey guys, I was using PKC on a Vero running OSMC right aroung when this issue was raised. I assume that I was experiencing the same as @bkury and opened a support [post](https://discourse.osmc.tv/t/plexkodiconnect-with-osmc/92165) on the OSMC forums. Can either of you guys confirm, that what I described there is the same as @bkury is describing here?

**Comment by croneter** [comment]
Jupp, seems to be the same issue, @iKeule257

**Comment by iKeule257** [comment]
I know you initially said that there is nothing to be "fixed", but as you reopened this issue, is there any code change to be expected?

**Comment by iKeule257** [comment]
@bkury Could you explain how I can make these changes locally after having PKC installed? I can't find `/plexbmchelper/subscribers.py` in `Kodi/addons/plugin.video.plexkodiconnect/resources/lib`. Unfortunately I got very limited python experience, so an explanation would be nice.

**Comment by bkury** [comment]
I assume that you are already on KODI Matrix/Nexus (python3 branch of PKC) ?

This fix is for KODI Leia (python2 branch of PKC), haven't looked at the new python3 re-write yet. 
I have to setup a test environment for KODI Nexus anyway, so I'll take a look as soon as I have some spare time.

But I still hope that @croneter is going to fix this obvious and in some rare circumstances even serious issue in "master".

**Comment by iKeule257** [comment]
> I assume that you are already on KODI Matrix/Nexus (python3 branch of PKC) ?

Yes I am

**Comment by bkury** [comment]
Just tested it: 
I don't have any problems running KODI Nexus with PKC 3.6.12 / 3.6.13

Are you sure that you are running the latest PKC version? Also, is it possible that not all your clients are KODI Matrix+ ?

**Comment by iKeule257** [comment]
I am running an up to date OSMC (Kodi 19.4) and there are still a lot of network requests as I described [here](https://discourse.osmc.tv/t/plexkodiconnect-with-osmc/92165)
But you mentioned in your initial post that it only happens when you are playing a file. For me these many requests happen 24/7, even if I'm not playing anything.

**Comment by bkury** [comment]
And your PKC version is also up2date? (3.6.12/3.6.13)

Yes, my problem only happens on KODI Leia when playing a file or when it's paused. (on timeline update).
So, if you are sure that all your clients are KODI Matrix+ and all have the newest verison of PKC 3.6.12+ then you might have another issue.

**Comment by iKeule257** [comment]
In fact I did stop using PKC because of this, but I installed it today on my 2 Vero devices and the amount of requests is immediately very high again. But I did install Kodi on my Windows PC earlier today to test if it's also affected and as far as I could see my Windows PC was not affected, but I don't know that's just because the Plex server is also running on that same PC. I will have to test another PC.

Using PKC 3.6.12

**Comment by TheEngineerGuy** [comment]
Not sure if this an issue with Coreelec 22, but the challenge is, the TTL for plex.direct DNS records is 6 hours, but what I see is a DNS request every 5 seconds. 

This means we are requesting DNS record query every 5 seconds even when the record clearly says that there is no need to refresh it for 6 hours. This is getting worse than the Samsung soundbar spam.

Is there a reason why TTL is not being respected?

<img width="1212" height="819" alt="Image" src="https://github.com/user-attachments/assets/591d7397-1c61-4c93-8104-0a87f7419836" />

```------------
Server:  dns.google
Address:  8.8.8.8

------------
Got answer:
    HEADER:
        opcode = QUERY, id = 2, rcode = NOERROR
        header flags:  response, want recursion, recursion avail.
        questions = 1,  answers = 1,  authority records = 0,  additional = 0

    QUESTIONS:
        xxxxxxxxxx.yyyyyyyyyyyyyyyy.plex.direct, type = A, class = IN
    ANSWERS:
    ->  xxxxxxxxxx.yyyyyyyyyyyyyyyy.plex.direct
        internet address = a.b.c.d
        ttl = 21600 (6 hours)

------------
